Core/MIPS/MIPSVFPUUtils.cpp

```diff
@@ -946,34 +936,182 @@ void vfpu_sincos_single(float angle, float &sine, float &cosine) {
 	}
 }
 
-float vfpu_sin_double(float angle) {
-	return (float)sin((double)angle * M_PI_2);
+float vfpu_sin_mod2(float a) {
+	float2int val;
+	val.f = a;
+
+	int32_t k = get_uexp(val.i);
+	if (k == 255) {
+		val.i = (val.i & 0xFF800001) | 1;
+		return val.f;
+	}
+
+	if (k < 0x68) {
+		val.i &= 0x80000000;
+		return val.f;
+	}
+
+	// Okay, now modulus by 4 to begin with (identical wave every 4.)
+	int32_t mantissa = get_mant(val.i);
+	if (k > 0x80) {
+		const uint8_t over = k & 0x1F;
+		mantissa = (mantissa << over) & 0x00FFFFFF;
+		k = 0x80;
+	}
+	// This subtracts off the 2.  If we do, flip sign to inverse the wave.
+	if (k == 0x80 && mantissa >= (1 << 23)) {
+		val.i ^= 0x80000000;
+		mantissa -= 1 << 23;
+	}
```

Comment:
Couldn't we just add a phase shift here for one of [sin, cos], and share the rest of the code between them?

Reply:
I almost had it shared, and still waffling on it.  But the NAN handling is a bit different and I was worried about the -0 cases making it messy.

-[Unknown]

Reply:
We definitely can swap sin/cos and negative though (and have to for CORDIC), I was just trying to minimize perf impact here.

-[Unknown]